R/derive_advs_params.R

```diff
@@ -0,0 +1,600 @@
+#'   within each parameter, an error will be thrown.
+#'
+#'   Additionally, if the input parameters are measured in different units but are mutually
+#'   convertible (e.g., centimeters for one parameter and inches for another), an automatic
+#'   conversion will be performed in order to uniform the values before calculating the ratio.
+#'
+#'   **Note:** Conversion factors comes from unit definitions as per the NCI Thesaurus
+#'
+#'   *Permitted Values:* A variable of the input dataset or a function call
+#'
+#' @inheritParams admiral::derive_param_bmi
+#'
+#' @details
+#' The analysis value of the new parameter is derived as
+#' \deqn{RATIO = \frac{NUMERATOR}{DENOMINATOR}}
+#'
+#'
+#' @return The input dataset with the new parameter added. Note, a variable will only
+#'    be populated in the new parameter rows if it is specified in \code{by_vars}.
+#'
+#' @family internal
+#' @keywords internal
+derive_param_ratio <- function(dataset,
+                               by_vars,
+                               numerator_code,
+                               denominator_code,
+                               set_values_to,
+                               constant_numerator = FALSE,
+                               constant_denominator = FALSE,
+                               filter = NULL,
+                               constant_by_vars = NULL,
+                               get_unit_expr = NULL) {
+  assert_vars(by_vars)
+  assert_data_frame(dataset, required_vars = exprs(!!!by_vars, PARAMCD, AVAL))
+  assert_character_scalar(numerator_code)
+  assert_character_scalar(denominator_code)
+  assert_varval_list(set_values_to, required_elements = "PARAMCD")
+  assert_param_does_not_exist(dataset, set_values_to$PARAMCD)
+  assert_logical_scalar(constant_numerator)
+  assert_logical_scalar(constant_denominator)
+  filter <- assert_filter_cond(enexpr(filter), optional = TRUE)
+  assert_vars(constant_by_vars, optional = TRUE)
+  get_unit_expr <- assert_expr(enexpr(get_unit_expr), optional = TRUE)
+
+  ### Default formula with no units conversion applied ----
+
+  ratio_formula <- expr(
+    !!sym(paste0("AVAL.", numerator_code)) /
+      !!sym(paste0("AVAL.", denominator_code))
+  )
+
+  ### If `get_unit_expr` provided then check units and enable units conversion ----
+
+  if (!missing(get_unit_expr) && !is.null(get_unit_expr)) {
+    param_units <- list()
+
+    # Check if units are the same within each of two parameters
+
+    for (param in c(numerator_code, denominator_code)) {
+      units_found <- dataset %>%
+        mutate(`_unit` = !!get_unit_expr) %>%
+        filter(PARAMCD == param) %>%
+        pull(`_unit`) %>%
+        unique()
+
+      if (length(units_found) != 1L) {
+        cli_abort(
+          "Multiple units {.val {units_found}} found for {.val {param}}.
+          Please review and update the units."
+        )
+      }
+
+      param_units[[param]] <- units_found
+    }
```

Comment:
```suggestion
units_found <- dataset %>%
    filter(PARAMCD %in% c(numerator_code, denominator_code)) %>%
    group_by(PARAMCD) %>%
    mutate(`_unit` = !!get_unit_expr) |> 
    summarise(`_distinct_units` = n_distinct(`_unit`))
    
    # Check / fail
    if(any(units_found$`_distinct_units`) > 1) 
    {
        cli_abort(
          "Multiple units {.val {units_found}} found for {.val {param}}. # Need to change this a bit I guess
          Please review and update the units."
        )
      }
```

Reply:
What's the rational for this change?

Reply:
Loops are difficult to debug and maintain if anything ever goes wrong. I would say that loops should only be used if we want some side effects (i.e. saving a file), we need to extract names to use somewhere else, or we don't know the output length (i.e. when there is some randomness somewhere). 

In this case, we will always have two parameters where we want to figure out if the unit is distinct within each param. I would argue that it is much easier to decode that functionality the non-looped code.

There are perhaps also some performance benefits to not looping, however, they would probably be tiny in this example.

However, this might be moot, as I think that we should do this functionality in a assertion. Either our own, or from admiraldev.

Reply:
No loops are there anymore.